repl: Customizable subprompt for multiline input

pckrishnadas88 opened the pull request:
**Description**:
This PR adds the ability to customise the REPL subprompt for multiline input. Previously, the REPL always used the fixed | prompt for multiline statements. With this change, the subprompt can be customised. Once feature finalised doc can be updated.

**Feature**: Customisable subprompt for multiline input.
**Fixes** : https://github.com/nodejs/node/issues/59401

```js
// custom-repl-subprompt.js
'use strict';
const repl = require('repl').start({ prompt: '> ' });

// Default behavior
// > const foo = {
// | bar: 42,
// | baz: 100
// | }
// > foo
// { bar: 42, baz: 100 }

// Set a custom multiline prompt
repl.setMultilinePrompt('... ');

// Example with custom prompt
// > const obj = {
// ... x: 1,
// ... y: 2
// ... }
// > obj
// { x: 1, y: 2 }

// Set an emoji as multiline prompt (UTF-8)
repl.setMultilinePrompt('😎 ');

// Example with emoji prompt
// > const emojiObj = {
// 😎 mood: 'happy',
// 😎 energy: 'high'
// 😎 }
// > emojiObj
// { mood: 'happy', energy: 'high' }

```
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

pckrishnadas88 committed:
repl: add customizable subprompt for multiline input

Add option to customize the REPL subprompt for multiline input.

BridgeAR commented on lib/internal/readline/interface.js:
```diff
@@ -415,6 +416,23 @@ class Interface extends InterfaceConstructor {
+   * Returns the current multiline prompt.
+   * @returns {string}
+   */
+  getMultilinePrompt() {
```
This is adding these methods to the readline interface, while we currently only use this in the REPL. I don't think we should regularly expose this API.

Since we already have the symbol anyway, I think we can just use those and directly set the symbol and remove this getter and setter.

BridgeAR commented on lib/internal/repl.js:
```diff
@@ -22,6 +22,7 @@ function createRepl(env, opts, cb) {
     ignoreUndefined: false,
     useGlobal: true,
     breakEvalOnSigint: true,
+    multilinePrompt: opts?.multilinePrompt ?? '| ',
```
I believe this option is currently not used at all? I do think it would in fact be nicer if we used the repl options for defining the value. So we should probably do that?

pckrishnadas88 replied:
Hi I'm currently down with fever please give me a day or two. Usually where we discuss and finalize the requirements? I will make the changes soon.

BridgeAR replied:
Get well soon! 🫖 
Discussing things in the PR is fine.

pckrishnadas88 replied:
@BridgeAR I’ve updated the PR. My branch was behind main by over 200 commits due to a 2–3 week gap, and I got a bit lost on this issue. Could you please review the code and confirm if this aligns with your recommendations? Some tests are still failing, likely because I’m not entirely sure the rebase was completed correctly after encountering conflicts.

BridgeAR reviewed (commented):
Thanks for opening the PR. I think we should not add anything to readline, but we could add the option.

I also wonder if we just want to set that option to '| ' when we start the internal REPL and not do that for any other REPL instances. That way we would have the default behavior that we want and unbreak external REPLs, while still allowing them to opt-in to the other behavior, when appropriate.

pckrishnadas88 committed:
repl: support customizable multiline subprompt

Add a  option to the REPL to allow custom prompts
for multiline input, replacing the default

pckrishnadas88 committed:
repl: support customizable multiline subprompt

lint fixes for repl file.

pckrishnadas88 committed:
Merge branch 'main' into feature/custom-repl-subprompt

pckrishnadas88 committed:
repl: fix ESLint errors in repl module
